[Driver] Load the standard library before starting parallel frontend invocations

harlanhaskins opened the pull request:
Add a new action, `LoadModuleJobAction`, that the driver can use to schedule a load of a given module before we fan out and invoke the frontend multiple times. This gives the module interface loader a chance to compile it from a module interface before we start with parallel invocations, avoiding starting potentially dozens of redundant compiles of a large module. Start by using this on the standard library.

Quick fix for rdar://52839445

harlanhaskins commented:
This still needs tests.

harlanhaskins commented on lib/Driver/ToolChains.cpp:
```diff
@@ -1109,6 +1109,40 @@ ToolChain::constructInvocation(const StaticLinkJobAction &job,
+                               const JobContext &context) const {
+
+  // Write a temporary file that imports the module we're trying to pre-load.
+  auto tempFile = context.getTemporaryFilePath("tmp.swift");
```
I'll probably give this a better name.

jrose-apple commented on include/swift/Driver/Action.h:
```diff
@@ -338,6 +339,24 @@ class StaticLinkJobAction : public JobAction {
+/// actions.
+class LoadModuleJobAction : public JobAction {
+  virtual void anchor();
+  StringRef moduleName;
```
This string should probably be owned by the job.

jrose-apple commented on lib/Driver/ToolChains.cpp:
```diff
@@ -1109,6 +1109,40 @@ ToolChain::constructInvocation(const StaticLinkJobAction &job,
+  llvm::raw_fd_ostream os(tempFile, err);
+  llvm::SmallString<32> importDecl("import ");
+  importDecl.append(job.getModuleName());
+  os << importDecl.str() << "\n";
```
It's not really okay for the toolchain to write files during invocation construction. But we do have `-import-module Foo` as a hidden frontend option anyway (because Playgrounds uses it), so maybe we can just pass an empty file instead.

jrose-apple commented on lib/Driver/Driver.cpp:
```diff
@@ -1749,6 +1749,16 @@ void Driver::buildActions(SmallVectorImpl<const Action *> &TopLevelActions,
+    if (!Args.hasArg(options::OPT_parse_stdlib)) {
+      auto loadModule = C.createAction<LoadModuleJobAction>(STDLIB_NAME);
+      TopLevelActions.push_back(loadModule);
+    }
```
If nothing else depends on it, it won't necessarily happen first. (Check out the tests using `-driver-print-bindings` for testing dependency info.)

jrose-apple commented on lib/Driver/Driver.cpp:
```diff
@@ -1731,6 +1731,25 @@ Driver::computeCompilerMode(const DerivedArgList &Args,
 
+/// Determines whether the given set of inputs has multiple .swift, .sil, or
+/// .sib inputs, which will require loading the standard library.
+static unsigned hasMultipleSourceFileInputs(ArrayRef<InputPair> inputs) {
```
`bool`?

harlanhaskins replied:
🤦‍♂

jrose-apple commented on lib/Driver/Driver.cpp:
```diff
@@ -1731,6 +1731,25 @@ Driver::computeCompilerMode(const DerivedArgList &Args,
+    case file_types::TY_SIB:
+      numSourceFileInputs += 1;
+      if (numSourceFileInputs > 1)
+        return true;
```
…I think you just refactored this and didn't fix it.

harlanhaskins commented on test/Driver/Dependencies/bindings-build-record-options.swift:
```diff
@@ -5,21 +5,21 @@
-// MUST-EXEC-INITIAL: inputs: ["./main.swift"], output: {{[{].*[}]}}, condition: run-without-cascading
-// MUST-EXEC-INITIAL: inputs: ["./other.swift"], output: {{[{].*[}]}}, condition: run-without-cascading
-// MUST-EXEC-INITIAL: inputs: ["./yet-another.swift"], output: {{[{].*[}]}}, condition: run-without-cascading
+// MUST-EXEC-INITIAL: inputs: ["./main.swift", ""], output: {{[{].*[}]}}, condition: run-without-cascading
```
These are unfortunate, the `load-module` action has empty strings for its outputs. Should these print as `""`, or should they be omitted?

jrose-apple replied:
I think it's fine; it's a debugging option and we'll live with it.

jrose-apple commented on lib/Driver/ToolChains.cpp:
```diff
@@ -1113,6 +1113,40 @@ ToolChain::constructInvocation(const StaticLinkJobAction &job,
+
+  // Create an empty temporary file to typecheck.
+  auto tempFile = context.getTemporaryFilePath("tmp.swift");
+  Arguments.push_back(context.Args.MakeArgString(tempFile));
```
Cute, I didn't think about just *making* a temporary file being good enough. I do suggest giving it a slightly nicer name with `getTemporaryFilePath("tmp", ".swift")`.

jrose-apple commented:
Looks good other than the weird helper function.

harlanhaskins committed:
[Driver] Load the standard library before starting parallel frontend invocations

Add a new action, LoadModuleJobAction, that the driver can use to schedule a
load of a given module before we fan out and invoke the frontend multiple
times. This gives the module interface loader a chance to compile it from a
module interface before we start with parallel invocations, avoiding starting
potentially dozens of redundant compiles of a large module. Start by using this
on the standard library.

Quick fix for rdar://52839445

harlanhaskins commented:
@swift-ci please test

harlanhaskins commented:
@swift-ci please python lint

harlanhaskins commented:
@swift-ci please python lint

harlanhaskins commented:
@swift-ci please test

harlanhaskins commented:
@swift-ci please test
